File: sphere.cpp
#include "sphere.hpp"

#include <cassert>
#include <cmath>

sphere::sphere(vec3 const& centre, double radius)
    : centre(centre), radius(radius) {}

std::optional<hit_record> sphere::hit(ray const& r, double const t_min, double const t_max) const {
    assert(t_min <= t_max);
    
    vec3 const oc = r.origin() - centre;
    double const
        a = r.direction().dot(r.direction()),
        b = oc.dot(r.direction()),
        c = oc.dot(oc) - radius * radius,
        discriminant = b * b - a * c,
        discriminant_root = std::sqrt(discriminant);

    if (discriminant > 0) {
        double const first_hit = (-b - discriminant_root) / a;
        if (first_hit < t_max && first_hit > t_min) {
            auto const hit = r.point_at_parameter(first_hit);
            return {{
                first_hit,
                hit,
                (hit - centre) / radius
            }};
        }

        double const second_hit = (-b + discriminant_root) / a;
        if (second_hit < t_max && second_hit > t_min) {
            auto const hit = r.point_at_parameter(second_hit);
            return {{
                second_hit,
                hit,
                (hit - centre) / radius
            }};
        }
    }

    return {};
}

File: rrt/src/sphere.rs
use crate::material::Material;
use crate::pt3::Pt3;
use crate::ray::Ray;
use crate::scene_object::{HitRecord, SceneObject};

pub struct Sphere<'a> {
    centre: Pt3,
    radius: f32,
    material: Box<dyn 'a + Material>,
}

impl<'a> Sphere<'a> {
    pub fn new<T: 'a + Material>(centre: Pt3, radius: f32, material: T) -> Self {
        Sphere {
            centre,
            radius,
            material: Box::new(material),
        }
    }
}

impl<'a> SceneObject<'a> for &'a Sphere<'a> {
    fn hit(self, ray: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord<'a>> {
        let oc = ray.origin - self.centre;
        let a = ray.direction.dot(ray.direction);
        let b = oc.dot(ray.direction);
        let c = oc.dot(oc) - self.radius * self.radius;
        let discriminant = b * b - a * c;

        let try_intersection = |t: f32| {
            if t < t_max && t > t_min {
                let p = ray.point_at_parameter(t);
                Some(HitRecord {
                    t,
                    p,
                    normal: (p - self.centre) / self.radius,
                    material: &*self.material,
                })
            } else {
                None
            }
        };

        if discriminant > 0.0 {
            if let Some(t) = try_intersection((-b - (b * b - a * c).sqrt()) / a) {
                return Some(t);
            }
            if let Some(t) = try_intersection((-b + (b * b - a * c).sqrt()) / a) {
                return Some(t);
            }
        }

        None
    }
}
